### PR TITLE
[ParseableInterface] Hack: prefer module names over type names 

### PR DESCRIPTION
**Pull request by jrose-apple**
...in the parts of a .swiftinterface file that are pretty-printed.

The effect of this is to dodge a longstanding issue ([SR-898](https://bugs.swift.org/browse/SR-898)) where if you have a module and type with the same name, the compiler always prefers the type. This is a problem for swiftinterface files, which are pretty-printed to use qualified types everywhere (to minimize the chance of collisions when a library changes). We know that one of these types exists in the wild, too: XCTest, the base class of XCTestCase.

This does not fix SR-898; it only works around it in certain limited cases by making the *opposite* assumption: preferring a module name instead of a type name. We also considered having some special syntax that indicates that the first component of a qualified name is a module, but felt that that might actually be a feature of the language some day, and wouldn't want to end up with two syntaxes for the same thing (or worse, have subtly different semantics for the one used in swiftinterfaces, or have it conflict with something *else).*

Builds on #22830.

### CODE REVIEW

**Comment by jrose-apple** [comment]
@swift-ci Please test

File: lib/Sema/TypeCheckType.cpp [hunk 1]
```diff
@@ -1212,6 +1212,56 @@ static Type diagnoseUnknownType(TypeResolution resolution,
...
 
+static bool shouldPreferModulesForTopLevelIdentTypeComponent(
+    const DeclContext *DC, TypeResolutionOptions options) {
+
+  // FIXME: We have a problem when a module and a type have the same name; the
+  // underlying UnqualifiedLookup code assumes that the type is always what's
+  // meant, even when that might not be the case. Work around this for
+  // parseable interfaces by making the *opposite* assumption outside of
+  // expressions and function bodies.
+  const SourceFile *containingSF = DC->getParentSourceFile();
```
Comment (harlanhaskins) [inline lib/Sema/TypeCheckType.cpp]: Nitpick: Can you make it more explicit here that this is just for parseable interfaces?

**Review by harlanhaskins — APPROVED** [review]
Overall looks good, for an unfortunate hack. I'm curious why this affects the API digester, though. Shouldn't `shouldPrintModuleContext` remain unchanged in the API digester?

**Comment by jrose-apple** [comment]
The API digester changes are part of #22830, not something additional here. I asked Xi what he thought about that over there but he hasn't replied yet.

**Comment by harlanhaskins** [comment]
Ah, okay. Thanks!

File: lib/Sema/TypeCheckType.cpp [hunk 2]
```diff
@@ -1212,6 +1212,56 @@ static Type diagnoseUnknownType(TypeResolution resolution,
   return ErrorType::get(ctx);
 }
 
+static bool shouldPreferModulesForTopLevelIdentTypeComponent(
```
Comment (slavapestov) [inline lib/Sema/TypeCheckType.cpp]: Can this become a method on TypeResolutionOptions, if you added some more state to that class?

File: lib/Sema/TypeCheckType.cpp [hunk 3]
```diff
@@ -1212,6 +1212,56 @@ static Type diagnoseUnknownType(TypeResolution resolution,
...
+  if (!containingSF || containingSF->Kind != SourceFileKind::Interface)
+    return false;
+
+  // If this is not a local context, just check whether we're /in/ a local
+  // context.
+  if (!DC->isLocalContext())
+    return DC->getLocalContext() == nullptr;
+
+  // We still want to return true for function parameters and result types, as
+  // long as they're not in *another* local context.
```
Comment (slavapestov) [inline lib/Sema/TypeCheckType.cpp]: Inlinable functions cannot contain nested types so I think some of these checks could be simplified.

File: test/ParseableInterface/inlinable-function.swift [hunk 4]
```diff
@@ -4,7 +4,10 @@
 // RUN: %target-swift-frontend -emit-module -o /dev/null -merge-modules %t/Test.swiftmodule -disable-objc-attr-requires-foundation-module -emit-parseable-module-interface-path %t/TestFromModule.swiftinterface -module-name Test
 // RUN: %FileCheck %s --check-prefix FROMMODULE --check-prefix CHECK < %t/TestFromModule.swiftinterface
 
-// CHECK: public struct Foo : Hashable {
+// FIXME: These shouldn't be different, or we'll get different output from
+// WMO and non-WMO builds.
+// FROMSOURCE-LABEL: public struct Foo : Hashable {
+// FROMMODULE-LABEL: public struct Foo : Swift.Hashable {
```
Comment (slavapestov) [inline test/ParseableInterface/inlinable-function.swift]: What causes this?

**Review by slavapestov — APPROVED** [review]

File: test/ParseableInterface/inlinable-function.swift (hunk 4 above)
Comment (jrose-apple) [reply test/ParseableInterface/inlinable-function.swift]: I'm not sure, but that's part of #22830. This PR is just about the last commit.

File: lib/Sema/TypeCheckType.cpp (hunk 3 above)
Comment (jrose-apple) [reply lib/Sema/TypeCheckType.cpp]: I don't think it gets any simpler than what's here. Inlinable functions can still contain other functions.

Comment (jrose-apple) [reply lib/Sema/TypeCheckType.cpp]: (and typealiases)

File: lib/Sema/TypeCheckType.cpp (hunk 2 above)
Comment (jrose-apple) [reply lib/Sema/TypeCheckType.cpp]: Sure. Is that better, though?

Comment (slavapestov) [reply lib/Sema/TypeCheckType.cpp]: I think its better to encapsulate all the variations on lookup semantics there, yeah.

File: lib/Sema/TypeCheckType.cpp (hunk 3 above)
Comment (slavapestov) [reply lib/Sema/TypeCheckType.cpp]: When checking parameters and results are we in the DC of the function or are we in the parent DC?

**Comment by slavapestov** [comment]
I wonder if we can come up with a more universal rule here. If you have a single type component, like Foo, we treat it as a type. If you have Foo.Bar, we treat Foo has a module. It would be interesting to see if that passes source compat.

I'd much prefer a single rule over adding yet another *%(#%& flag to NameLookup...

File: lib/Sema/TypeCheckType.cpp (hunk 3 above)
Comment (jrose-apple) [reply lib/Sema/TypeCheckType.cpp]: The function, because we need to find the function's generic parameters.

**Comment by jrose-apple** [comment]
This rule would *definitely* break source compat. All it takes is a nested type in the type that has the same name as your module. (Also, I don't want to tie parseable interfaces to a language change.)

I share your distaste, though. I suspect the real answer will be that some day we'll have an actual syntax for module qualification and then this will only be legacy support.

**Comment by davidungar** [comment]
I just finished a refactoring of UnqualifiedLookup. I haven’t had a chance to look at the code yet, but I’ll chime in to agree with Jordan’s discomfort and Slava’s reservations. If there is any other way to meet the larger goals, it would be better to avoid more options in that code, which already has much that seems inexplicable or opaque, particularly at that module scope level. 

- David

**Comment by jrose-apple** [comment]
I haven't thought of anything that doesn't require inventing syntax, though, and that definitely seems worse (even less maintainable).

**Comment by davidungar** [comment]
@jrose-apple I've added myself as a reviewer, since I just refactored some of the name lookup code. I'll be happy to help figuring out how to make the change in the refactored world, assuming you go through with it. Does it make sense to split up the PR, so the nameLookup part goes in first?
Also, FYI, in a day, I'll be submitting a PR to remove the legacy-cross-check. The change in this PR will be easier to put in after that happens.

**Comment by jrose-apple** [comment]
Thanks! I have the PR split up the other way ("print everything qualified"), and that first part still has bugs in it, so I might not get back to this right away.

**Comment by davidungar** [comment]
My pleasure! Good luck!!

**Pull request closed without merging**